lib/sgtn-client/api/base.rb

```diff
@@ -0,0 +1,92 @@
+# 
+#  Copyright 2019-2022 VMware, Inc.
+#  SPDX-License-Identifier: EPL-2.0
+#
+
+require 'multi_json'
+
+module SgtnClient
+
+  class Base
+    class << self
+
+      def getTranslation(component, key, locale)
+        locale = SgtnClient::LocaleUtil.get_best_locale(locale)
+        items = get_cs(component, locale)
+        if items.nil? || items["messages"].nil?
+          nil
+        else
+          items["messages"][key]
+        end
+      end
+
+      def compareSource(component, key, default_locale, source, translation)
+        SgtnClient.logger.debug "[Base][compareSource]component=#{component},key=#{key},default_locale=#{default_locale},source=#{source},translation=#{translation}"
+        items = get_cs(component, default_locale)
+        if items.nil? || items["messages"].nil?
+          translation
+        else
+          target = items["messages"][key]
+          source == target ? translation : source
+        end
+      end
```

Comment:
seems you are saying component-based comparison, my current compareSource is for getString only, we could have another PR for component-based comparison.

Reply:
If you look at the design, the solution will cover both at once.

Reply:
I have concern on the component-based comparison, will handle it in other PR.

Reply:
Then we shall block this PR because according to your reasoning, it does not solve issue 1460 completely (error will still happen for get_strings). More than that, I won't aprrove this PR because of the followng reasons:

1. The source-comparison logic manipulates the DATA. Hence, it should be in the DATA layer. You put it in the client API endpoint (getString) instead which is not the place for it.
2. You plan to implement separately for getString and getStrings. However, source-comparison logic should not be duplicated. Source comparison is applied to a SINGLE data set (i.e. translation bundle). It is the same data set that you access regardless of whether the request is get_string or get_strings. 
3. CLIENT-side cached data should always be updated for the CLIENT. This means it should have the most recently updated source instead of an old translation. Client-side cache is not meant to be a mere replica of the service-side cache. 

The code suggestion I put in this PR handles BOTH cases (getString and getStrings) with FEWER lines of code. It correctly saves updated data in client-side cache.

Reply:
1460 should be break to two issues, one for getString, another for getStrings, how do you think? @Xiaochao8

Reply:
If this is urgent, we can fix getSting in this PR. But keep in mind, this is a temporaty solution and will be overrided by comparing component strings.

Reply:
> If you look at the design, the solution covers both at once. We don't need source comparison in getString. Refer to the design spec please. [L10n Fetch](https://confluence.eng.vmware.com/pages/viewpage.action?pageId=530637182#SingletonClientLibraryCaching,FallbackMechanism,andOtherFeatures-L10nFetch) and [Source Comparison](https://confluence.eng.vmware.com/pages/viewpage.action?pageId=530637182#SingletonClientLibraryCaching,FallbackMechanism,andOtherFeatures-SourceComparison)

I compared the spec with the earlier version(before Feb 24. 2022), there are much difference and the details of source comparison was not mentioned before, so i agree it's a solution to fix get_string. We do need agreement on source comparison part and upgrade it in next versions. More important shouldn't implement the spec without considering the risks bring to the integration product.

Reply:
The spec was not modified recently. It was carefully discussed way back in 2021, only the document was updated this week.